File: Review-06-Python-Shapes/Example-6/shapes/shape_factory.py
"""
This module provides factory utilities for creating shapes. This includes
recording which Shape types are available.

"""

import copy

from shapes.circle import Circle
from shapes.square import Square
from shapes.triangle import (Triangle, RightTriangle, EquilateralTriangle)


_KNOWN_SHAPES = {
    "Triangle": (
        Triangle(),
        lambda a, b, c: Triangle(a, b, c)
    ),
    "Right Triangle": (
        RightTriangle(),
        lambda base, height: RightTriangle(base, height)
    ),
    "Equilateral Triangle": (
        EquilateralTriangle(),
        lambda side: EquilateralTriangle(side)
    ),
    "Square": (
        Square(),
        lambda side: Square(side)
    ),
    "Circle": (
        Circle(),
        lambda radius: Circle(radius)
    )
}  # _Dictionary_ of known shapes


def create(name):
    """
    Create a Shape

    Args:
        name: the shape to be created

    Returns:
        A shape with the specified name or null if no matching shape is found
    """

    if name in _KNOWN_SHAPES:
        return copy.deepcopy(_KNOWN_SHAPES[name][0])

    return None


def create_from_dimensions(name, values):
    """
    Create a Shape

    Args:
        name: the shape to be created

        values: dictionary of values corresponding to the data needed
            to inialize a shape

    Returns:
        A shape with the specified name or null if no matching shape is found
    """

    if name in _KNOWN_SHAPES:
        return _KNOWN_SHAPES[name][1](*values)

    return None


def is_known(name):
    """
    Determine whether a given shape is known

    Args:
        name: the shape for which to query
    """

    return name in _KNOWN_SHAPES


def list_known():
    """
    Print a list of known Shapes
    """
    return "\n".join((f"  {name:}" for name in _KNOWN_SHAPES))


def number_known():
    """
    Determine the number of known Shapes
    """

    return len(_KNOWN_SHAPES)

File: Review-06-Rust-Shapes/Example-2/tests/shape_factory.rs
#[cfg(test)]
#[macro_use]
extern crate hamcrest2;
extern crate shapes;

use hamcrest2::prelude::*;

use shapes::factory::Factory;

#[test]
fn test_is_known() {
    let factory = Factory::new();

    assert!(factory.is_known("Circle"));
    assert!(factory.is_known("Square"));
    assert!(factory.is_known("Triangle"));
    assert!(factory.is_known("Right Triangle"));
    assert!(factory.is_known("Equilateral Triangle"));
}

#[test]
fn test_number_known() {
    let factory = Factory::new();

    assert_that!(factory.number_known(), is(equal_to(5)));
}

#[test]
fn test_str() {
    let factory = Factory::new();
    let f_str = factory.list_known();

    assert!(f_str.contains("  Circle"));
    assert!(f_str.contains("  Square"));
    assert!(f_str.contains("  Triangle"));
    assert!(f_str.contains("  Right Triangle"));
    assert!(f_str.contains("  Equilateral Triangle"));
}
